File: chapters/tree_traversal/code/julia/Tree.jl
using DataStructures

struct Node
    children::Vector{Node}
    ID::Int64
    Node(ID::Int64) = new(Vector{Node}(), ID)
end

function DFS_recursive(n::Node)
    # Here we are doing something...
    println(n.ID)

    for child in n.children
        DFS_recursive(child)
    end
end

function DFS_recursive_postorder(n::Node)

    for child in n.children
        DFS_recursive_postorder(child)
    end

    # Here we are doing something...
    println(n.ID)
end

# This assumes only 2 children
function DFS_recursive_inorder_btree(n::Node)

    if (length(n.children) > 2)
        println("Not a binary tree!")
        exit(1)
    end

    if (length(n.children) > 0)
        DFS_recursive_inorder_btree(n.children[0])
        println(n.ID)
        DFS_recursive_inorder_btree(n.children[1])
    else
        println(n.ID)
    end
end

function DFS_stack(n::Node)
    s = Stack(Node)
    push!(s, n)

    while(length(s) > 0)
        println(top(s).ID)
        temp = pop!(s)
        for child in temp.children
            push!(s, child)
        end
    end
end

function BFS_queue(n::Node)
    q = Queue(Node)
    enqueue!(q, n)

    while(length(q) > 0)
        println(front(q).ID)
        temp = dequeue!(q)
        for child in temp.children
            enqueue!(q, child)
        end
    end
end

# function to create a simple, balanced tree
function create_tree(num_row::Int64, num_child::Int64)
    ret = Node(num_row)
    if (num_row == 0)
        return ret
    end

    for i = 1:num_child
        child = create_tree(num_row - 1, num_child)
        push!(ret.children, child)
    end

    return ret
end

function main()

    println("Creating Tree")
    root = create_tree(2,3)

    println("Using recursive DFS:")
    DFS_recursive(root);

    println("Using stack-based DFS:")
    DFS_stack(root);

    println("Using queue-based BFS:")
    BFS_queue(root);
end

main()

File: chapters/tree_traversal/code/java/Tree.java
// submitted by xam4lor
import java.util.ArrayList;
import java.util.PriorityQueue;
import java.util.Queue;
import java.util.Stack;

public class Tree {
    // root node of the Tree
    public Node root;

    public Tree(int numberRow, int numberChild) {
        root = createTree(numberRow, numberChild);
    }



	 public void DFSRecursive(Node node) {
	     System.out.println(node.id);

	     for (Node n : node.children) {
	         DFSRecursive(n);
	     }
	 }

	 public void DFSRecursivePostOrder(Node node) {
	     for (Node n : node.children) {
	         DFSRecursivePostOrder(n);
	     }

	     // Here we are doing something ...
	     System.out.println(node.id);
	 }

	 // This assumes only 2 children
	 public void DFSRecursiveInOrderBTree(Node node) {
	     if(node.children.size() > 2) {
	         System.out.println("Not a binary tree!");
	         System.exit(1);
	     }

	     if(node.children.size() > 0) {
	         DFSRecursiveInOrderBTree(node.children.get(0));
	         System.out.println(node.id);
	         DFSRecursiveInOrderBTree(node.children.get(1));
	     }
	     else {
	         System.out.println(node.id);
	     }
	 }

	 public void DFSStack() {
	     Stack<Node> stack = new Stack<Node>();
	     stack.push(this.root);

	     Node tmp;

	     while(stack.size() != 0) {
	         System.out.println(stack.peek().id);
	         tmp = stack.pop();

	         for (Node c : tmp.children) {
	             stack.push(c);
	         }
	     }
	 }

	 public void BFSQueue() {
	     Queue<Node> queue = new PriorityQueue<Node>();
	     queue.add(this.root);

	     Node temp;

	     while(queue.size() != 0) {
	         System.out.println(queue.peek().id);
	         temp = queue.poll(); // return null if the queue is empty

	         if(temp != null) {
	             for (Node c : temp.children) {
	                 queue.add(c);
	             }
	         }
	     }
	 }




     public Node createTree(int numberRow, int numberChild) {
	     Node ret = new Node(numberRow);

	     if(numberRow == 0) {
	         return ret;
	     }

	     for (int i = 1; i < numberChild; i++) {
	         Node child = createTree(numberRow - 1, numberChild);
	         ret.children.add(child);
	     }

		 return ret;
	 }



     private class Node implements Comparable<Node> {
	     public ArrayList<Node> children;
	     public int id;

	     public Node(int id) {
	         this.children = new ArrayList<Node>();
	         this.id = id;
	     }

	     @Override
	     public int compareTo(Node other) {
	         // Need to implement Comparable<Node> and override this
	         // method because of the method BFSQueue() which uses Queues
	         // and must know how to check if two nodes are the same or not
	         return Integer.compare(this.id, other.id);
	     }
	 }
}
